ang/civicase/PanelQuery.js

```diff
@@ -0,0 +1,196 @@
+    $scope.title = '---';
+    $scope.total = 0;
+    $scope.selectedRange = 'week';
+    $scope.periodRange = [
+      { label: 'This Week', value: 'week' },
+      { label: 'This Month', value: 'month' }
+    ];
+    $scope.pagination = {
+      page: 1,
+      size: PAGE_SIZE,
+      range: { from: 1, to: PAGE_SIZE }
+    };
+
+    (function init () {
+      initWatchers();
+      verifyData();
+
+      loadData();
+    }());
+
+    /**
+     * Calculates the offset of the results list based
+     * on the current page number and page size
+     *
+     * @return {String}
+     */
+    function calculatePageOffset () {
+      return ($scope.pagination.page - 1) * $scope.pagination.size;
+    }
+
+    /**
+     * Fetches the data via the Civi API and stores the response
+     *
+     * @param {Boolean} skipCount if true then the "getcount" request won't be sent
+     * @return {Promise}
+     */
+    function fetchDataViaApi (skipCount) {
+      var apiCalls = {
+        get: [ $scope.query.entity, $scope.query.action || 'get', prepareRequestParams() ],
+        count: [ $scope.query.entity, 'getcount', $scope.query.params ]
+      };
+
+      skipCount && (delete apiCalls.count);
+
+      return crmApi(apiCalls)
+        .then(function (result) {
+          !skipCount && ($scope.total = result.count);
+
+          return processResults(result.get.values);
+        })
+        .then(function (processed) {
+          $scope.results = processed;
+        });
+    }
+
+    /**
+     * Initializes the directive's watchers
+     */
+    function initWatchers () {
+      // Triggers a refresh when the query params change
+      $scope.$watchCollection('query.params', function (newParams, oldParams) {
+        (newParams !== oldParams) && loadData();
+      }, true);
+
+      // Triggers the range handler (if present) when the selected range changes
+      $scope.$watch('selectedRange', function (newRange, oldRange) {
+        if (newRange !== oldRange && $scope.handlers.range) {
+          $scope.handlers.range($scope.selectedRange, $scope.query.params);
+        }
+      });
+
+      // Triggers a new request and a recalculation of the pagination range
+      // when the current page changes
+      $scope.$watch('pagination.page', function (newPage, oldPage) {
+        (newPage !== oldPage) && loadData(true);
+      });
+    }
+
+    /**
+     * Loads the data and triggers any subsequent logic
+     *
+     * @param {Boolean} skipCount sets whether the directive needs to recalculate the total
+     * @return {Promise}
+     */
+    function loadData (skipCount) {
+      if (!skipCount) {
+        $scope.pagination.page = 1;
+        $scope.loading = true;
+      }
+
+      return fetchDataViaApi(skipCount)
+        .then(updatePaginationRange)
+        .then(function () {
+          if (!skipCount) {
+            $scope.title = $scope.handlers.title ? $scope.handlers.title($scope.total) : $scope.title;
+            $scope.loading = false;
```

Comment:
move to `.finally` statement

Reply:
This is now a bit different, but I don't see the benefit of having it in `.finally` instead of `.then`. `.finally` will be called even if there are errors, while the loading state should be lifter only if the data fetching+processing is successful (unfortunately we don't have much spec'd out in terms of error states)

Reply:
wait, crmApi runs successfully even if it finds an error? 😮